### ISSUE TITLE
[LOW] [P2P] Allow for QR codes to be easily saved or copied on all platforms 

### ISSUE THREAD

**Issue by MitchExpensify** [Expensify/App#19834]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
On desktop:
1. Click any room header
2. Click "Share code"
3. Right click the QR code

## Expected Result:
Have the option to "Save image as", "Copy image", or Download the QR code image 

## Actual Result:
No option to easily save or copy the QR code image

## Workaround:
Copy or save the QR code image on mobile 

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** v1.3.19-7
**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes 
**Email or phone of affected tester (no customers):** All
**Notes/Photos/Videos:** 

<img width="1185" alt="image" src="https://github.com/Expensify/App/assets/36425901/60a0b258-d18a-4ebf-9d81-32583d29b348">

<img width="1206" alt="image" src="https://github.com/Expensify/App/assets/36425901/0d4d24a8-b95c-41f5-b488-1e92f75cf31d">

**Expensify/Expensify Issue URL:**
**Issue reported by:** @MitchExpensify 
**Slack conversation:** Internal convo: https://expensify.slack.com/archives/C03U7DCU4/p1685041725518029, #bug-zero convo tbd 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~013d2f9dd751374d8e</li>
        <li>Upwork Job ID: 1663602988972146688</li>
        <li>Last Price Increase: 2023-05-30</li>
    </ul>
</details>

**Comment by JmillsExpensify** [comment]
Great point reminder that we should always optimize for cross-platform everywhere!

**Comment by robertjchen** [comment]
@chrispader Would you possibly have some free cycles to look into adding QR code download for Web? I recall you mentioning that we may have to make some upstream changes to the library? 🤔

**Comment by chrispader** [comment]
Yes, i can take a look at that! :)

**Comment by chrispader** [comment]
Yes, so the `react-native-view-shot` library currently doesn't work with `react-native-web`, so i'm gonna investigate the problem and open an upstream PR. We could probably also open a separate PR in our an own fork?

**Comment by robertjchen** [comment]
yup, sounds like a plan! 🙇

**Comment by chrispader** [comment]
Working on [this issue](https://github.com/gre/react-native-view-shot/issues/224) in the upstream library

**Comment by chrispader** [comment]
The PR is working and making screenshots on web works now too, though there is one problem.

The library that is creating a canvas from the html canvas doesn't have access to the profile picture/avatar image that is received from CloudFront because of CORS policy regulations.

Is there any way to add [these CORS headers](https://stackoverflow.com/a/72430377/5774312) to the S3 Bucket/AWS or would this violate Expensify's security?

**Comment by chrispader** [comment]
Which leads to images like this:

![Christoph Pader-ShareCode (8)](https://github.com/Expensify/App/assets/20173411/9bc345eb-814c-4e08-bce3-fcd72d06d96d)

**Comment by chrispader** [comment]
The `Expensify/App` PR for this feature is [here](https://github.com/Expensify/App/pull/20356)

**Comment by robertjchen** [comment]
Yep, I'll look into the CORs changes!

**Comment by chrispader** [comment]
@robertjchen are there any updates on the CORS changes?

**Comment by robertjchen** [comment]
Nothing yet, will report back before EOW 👍

**Comment by robertjchen** [comment]
Apologies for the delay, some firefighting last week.

**Comment by robertjchen** [comment]
This is on me! I'm looking into this question:

> Is there any way to add [these CORS headers](https://stackoverflow.com/a/72430377/5774312) to the S3 Bucket/AWS or would this violate Expensify's security?

The straight answer is that we generally want to only allow requests from clients on Expensify domains (otherwise someone could potentially use us as an image hosting site by uploading profile photos and linking them elsewhere).

Don't think it's being hosted elsewhere, so it must be a local CORs issue... will need to dive a bit deeper 👀

**Comment by robertjchen** [comment]
> Which leads to images like this:
> 
> ![Christoph Pader-ShareCode (8)](https://user-images.githubusercontent.com/20173411/244051344-9bc345eb-814c-4e08-bce3-fcd72d06d96d.png)

@chrispader could you let me know if/what the CORs error is displayed in the browser console when making a screenshot? 🙏

**Comment by chrispader** [comment]
Hey @robertjchen!

There is no error displayed in the browser's console, because the`html2canvas` library internally just isn't able to retrieve any images from a different domain.

> html2canvas does not get around content policy restrictions set by your browser. Drawing images that reside outside of the origin of the current page taint the canvas that they are drawn upon. If the canvas gets tainted, it cannot be read anymore. If you wish to load images that reside outside of your pages origin, you can use a proxy to load the images.

https://html2canvas.hertzen.com/proxy

**Comment by chrispader** [comment]
That's why images from different domains (such as `d1wpcgnaa73g0y.cloudfront.net` for my profile picture)cannot be rendered and therefore are not in the screenshot...

![Screenshot 2023-07-09 at 09 02 15](https://github.com/Expensify/App/assets/20173411/1d1e2cce-d917-4ed3-bfba-f8d21bc72f67)

**Comment by chrispader** [comment]
To have the profile pictures in the screenshot, we would need to create a proxy that runs inside the "Expensify" Domain, just to retrieve the images from the same "origin"/domain. I guess this would be a way to big overhead, right?

[`html2canvas`has a `useCORS` config option](https://html2canvas.hertzen.com/configuration), that's why i though we might be able to display the images anyway, if the CORS headers are valid.

**Comment by greg-schroeder** [comment]
What's the status on this one @robertjchen?

**Comment by robertjchen** [comment]
Created an internal proposal for infrastructure changes, will report back after that's done.

**Comment by chrispader** [comment]
@robertjchen from my understanding, the profile pics would also need to be under `new.expensify.com`. `profiles.expensify.com` would already be a different origin/domain, see [here](https://en.wikipedia.org/wiki/Same-origin_policy#Origin_determination_rules)

**Comment by robertjchen** [comment]
Yep, but I think we can then relax the origin policy to allow for anything `*.expensify.com`  rather than `*` 🤔

**Comment by chrispader** [comment]
hmm, not 100% sure, but yea i think that should be possible 👍

**Comment by chrispader** [comment]
Where is the origin policy defined?

**Comment by chrispader** [comment]
Just tested it with the new domain, but it still doesn't work. I suppose, we'll still have to add the `Access-Control-Allow-Origin` header with `*.expensify.com`?

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin

**Comment by robertjchen** [comment]
Ah, you were right [originally](https://github.com/Expensify/App/issues/19834#issuecomment-1639835509)- looks like `*.expensify.com` [isn't allowed](https://portswigger.net/web-security/cors/access-control-allow-origin#:~:text=Note%20that%20wildcards%20cannot%20be%20used%20within%20any%20other%20value) 🤔 

It seems like we have to send the proper CORs headers on-the-fly based on the origin of the request. I'm not sure if we can get these profile pics onto the `new.expensify.com` domain without some major changes (which might be more work than doing the former).

**Comment by greg-schroeder** [comment]
Okay where did we land on this one @robertjchen?

**Comment by robertjchen** [comment]
On vacay, but will try to figure out the remaining hurdle after I get back on Wednesday. We're halfway there!

**Comment by greg-schroeder** [comment]
Sounds good!

**Comment by robertjchen** [comment]
Coming up with a plan and getting it buddy-checked, since what I'm planning on doing has the potential to break all of NewDot Web if the settings are right.

**Comment by robertjchen** [comment]
Plan to be reviewed in the corresponding internal issue: #300566 since it contains backend details that should not be public.

**Comment by greg-schroeder** [comment]
Thanks @robertjchen!

**Comment by robertjchen** [comment]
Getting proposal reviewed before I break all of NewDot, should have the changes applied by EOW

**Comment by greg-schroeder** [comment]
Is there a place I can follow along on that discussion/proposal @robertjchen ?

**Comment by robertjchen** [comment]
Yep, check the issue # from my comment here: https://github.com/Expensify/App/issues/19834#issuecomment-1682028722

**Comment by greg-schroeder** [comment]
Where do we stand here @robertjchen? <3

**Comment by shawnborton** [comment]
What's the latest here?

**Comment by robertjchen** [comment]
Getting backend changes reviewed while we're all here during Offshore.

**Comment by greg-schroeder** [comment]
Progress continues

**Comment by greg-schroeder** [comment]
Offshore is over! Thoughts on this one @robertjchen?

**Comment by robertjchen** [comment]
Got the green light 🟢 for the backend changes- going to do it later this week when things are more stable 🤞

**Comment by greg-schroeder** [comment]
Maybe this week's the week? :) @robertjchen

**Comment by robertjchen** [comment]
two fires so far, hoping not to add wood to it 😄

**Comment by robertjchen** [comment]
Moving to `Monthly` for proper expectations

**Comment by greg-schroeder** [comment]
Seems like no update here

**Comment by greg-schroeder** [comment]
Same as above

**Comment by greg-schroeder** [comment]
@robertjchen would you say this fits into a current wave or VIP? If not and there hasn't been any more progress, perhaps this can be closed.

**Comment by robertjchen** [comment]
This still needs to be done- the final steps should be pretty straightforward. We can file it under VIP starting next year if it's not completed before then 👍

**Comment by arielgreen** [comment]
Adding this to the Group release under #vip-splits.

**Comment by Christinadobrzyn** [comment]
I'm closing this [GH](https://github.com/Expensify/App/issues/33815) as a dupe of this issue. Feel free to let me know if you think they need to be treated separately.

**Comment by greg-schroeder** [comment]
This has been sitting forever - sounds like we classified it as a nice to have? Should this still be listed as Critical or even open at all?

**Comment by robertjchen** [comment]
Changed to `Low - P2P` per https://expensify.slack.com/archives/C05RECHFBEW/p1705954738163119?thread_ts=1705712907.222529&cid=C05RECHFBEW

**Comment by robertjchen** [comment]
Also updating assignments for cleaner K2s and since this is now an `Internal` task.

**Comment by robertjchen** [comment]
Now prioritized under #vip-split

**Comment by robertjchen** [comment]
Revisiting this later this week!

**Comment by robertjchen** [comment]
Was focused on pushing out wave items, revisiting this this week for sure!

**Comment by robertjchen** [comment]
Tackling this week, moved https://github.com/Expensify/Expensify/issues/300566 to weekly

**Comment by robertjchen** [comment]
https://github.com/Expensify/Expensify/issues/300566 to be addressed frst

**Comment by robertjchen** [comment]
paused for now

**Comment by robertjchen** [comment]
no update

**Comment by robertjchen** [comment]
no update

**Comment by robertjchen** [comment]
no update

**Comment by robertjchen** [comment]
no update

**Comment by robertjchen** [comment]
no update

**Comment by robertjchen** [comment]
on hold

**Comment by robertjchen** [comment]
not a priority at the moment

**Comment by robertjchen** [comment]
not a priority at the moment

**Comment by robertjchen** [comment]
not a priority at the moment

**Comment by robertjchen** [comment]
not a priority at the moment

**Comment by robertjchen** [comment]
No update

**Comment by robertjchen** [comment]
no update

**Comment by robertjchen** [comment]
movement on backend portion
